File: Data Structures & Algorithms/balanced-binary-tree/submission-3.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isBalanced(self, root: Optional[TreeNode]) -> bool:
        def dfs(root):
            if not root:
                return [True, 0]
            left = dfs(root.left)
            right = dfs(root.right)
            if not left[0] or not right[0] or abs(left[1]-right[1]) > 1:
                return [False, 1 + max(left[1], right[1])]
            return [True, 1 + max(left[1], right[1])]
        return dfs(root)[0] 
        
        
    # def height(self, node: TreeNode) -> int:
    #     if not node:
    #         return 0
    #     lh = self.height(node.left)
    #     rh = self.height(node.right)
    #     return 1 + max(lh, rh)

File: Data Structures & Algorithms/trapping-rain-water/submission-3.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n = height.size();
        int l = 0, r = n-1;
        int leftMax = height[l], rightMax = height[r];
        int ans = 0;
        while (l < r) {
            if (leftMax < rightMax){
                l++;
                leftMax = max(leftMax, height[l]);
                ans += max(0, leftMax - height[l]);
            } else {
                r--;
                rightMax = max(rightMax, height[r]);
                ans += max(0, rightMax - height[r]);
            }
        }
        return ans;
    }
};
